riscv_executor/src/lib.rs

```diff
@@ -274,10 +285,19 @@ mod builder {
         /// advance to next row, returns the index to the statement that must be
         /// executed now, or None if the execution is finished
         pub fn advance(&mut self, was_nop: bool) -> Option<u32> {
+            if !was_nop {
+                self.trace.len += 1;
+            }
+
             if self.get_reg_idx(self.pc_idx) != self.get_reg_idx_next(self.pc_idx) {
-                // PC changed, create a new line
-                self.curr_idx += self.reg_len();
-                self.trace.regs.extend_from_within(self.curr_idx..);
+                if let ExecMode::Trace = self.mode {
+                    // PC changed, create a new line
+                    self.curr_idx += self.reg_len();
+                    self.trace.regs.extend_from_within(self.curr_idx..);
+                } else if !was_nop {
+                    let next_idx = self.curr_idx + self.reg_len();
+                    self.trace.regs.copy_within(next_idx.., self.curr_idx);
+                }
```

Comment:
This block here is not the best, but it does the job since `self.curr_idx` does not grow, so the vector doesn't grow either. This works for now and can probably be optimized once @lvella is back